docker/efficiency/builds.py

```diff
@@ -0,0 +1,221 @@
+import json
+import os
+import re
+import six
+import tarfile
+import collections
+
+from . import commons
+from .. import errors
+
+
+build_success_re = r'^Successfully built ([a-f0-9]+)\n$'
+
+# these prefixes are treated as remote by the docker daemon
+# (ref: pkg/urlutil/*) as of v1.6.0
+REMOTE_CONTEXT_PREFIXES = ["http://",
+                           "https://",
+                           "git://",
+                           "git@",
+                           "github.com/"]
+
+
+def get_build_id(build_result, discard_logs=False):
+    """ **Params:**
+        * `build_result` is a python generator returned by `Client.build`
+        * `discard_logs` (bool, default=False). If True, log lines will
+          be discarded after they're processed. Limits memory footprint.
+        **Returns** tuple:
+            1. Image ID if found, None otherwise
+            2. List of log lines
+    """
+    parsed_lines = []
+    image_id = None
+    for line in build_result:
+        try:
+            parsed_line = json.loads(line).get('stream', '')
+            if not discard_logs:
+                parsed_lines.append(parsed_line)
+            match = re.match(build_success_re, line)
+            if match:
+                image_id = match.group(1)
+        except ValueError:
+            # sometimes all the data is sent on a single line
+            # This ONLY works because every line is formatted as
+            # {"stream": STRING}
+            lines = re.findall('{\s*"stream"\s*:\s*"[^"]*"\s*}', line)
+            return get_build_id(lines, discard_logs)
```

Comment:
This sounds like a bug, do you know we can reproduce it?

If there is an open issue on engine, maybe add the url to the issue here